vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+
+```r
+## Definition of the time-step and output as "time"
+dt <- user(1)
+initial(time) <- 0
+update(time) <- (step + 1) * dt
+
+## Core equations for transitions between compartments:
+update(S[]) <- S[i] - n_SI[i]
+update(I[]) <- I[i] + n_SI[i] - n_IR[i]
+update(R[]) <- R[i] + n_IR[i]
+
+## Individual probabilities of transition:
+p_SI[] <- 1 - exp(-lambda[i] * dt) # S to I
+p_IR <- 1 - exp(-gamma * dt) # I to R
+
+## Force of infection
+m[, ] <- user() # age-structured contact matrix
+s_ij[, ] <- m[i, j] * I[i]
+lambda[] <- beta / N * sum(s_ij[i, ])
+
+## Draws from binomial distributions for numbers changing between
+## compartments:
+n_SI[] <- rbinom(S[i], p_SI[i])
+n_IR[] <- rbinom(I[i], p_IR)
+
+## Total population size
+N <- sum(S) + sum(I) + sum(R)
+
+## Initial states:
+initial(S[]) <- S_ini
+initial(I[]) <- I_ini
+initial(R[]) <- 0
+
+## User defined parameters - default in parentheses:
+S_ini <- user(1000)
+I_ini <- user()
+beta <- user(0.2)
+gamma <- user(0.1)
+
+# dimensions of arrays
+N_age <- user()
+dim(S) <- N_age
+dim(I) <- N_age
+dim(R) <- N_age
+dim(n_SI) <- N_age
+dim(n_IR) <- N_age
+dim(p_SI) <- N_age
+dim(m) <- c(N_age, N_age)
+dim(s_ij) <- c(N_age, N_age)
+dim(lambda) <- N_age
+```
+
+As before, save the file, and use `odin.dust` to compile the model:
+
+```{r}
+gen_age <- odin_dust("sirage.txt")
+```
+
+We can generate an age-structured contact matrix based on the [POLYMOD](https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.0050074) survey by using the [socialmixr](https://github.com/sbfnk/socialmixr) package:
+
+```{r}
+library(socialmixr)
+data(polymod)
+
+age.limits = seq(0, 70, 10)
+
+## Get the contact matrix from socialmixr
+contact <- socialmixr::contact_matrix(
+  survey = polymod,
+  countries = "United Kingdom",
+  age.limits = age.limits,
+  symmetric = TRUE)
+
+## Transform the matrix to the (symetrical) transmission matrix
+## rather than the contact matrix
+transmission <- contact$matrix /
+  rep(contact$demography$population, each = ncol(contact$matrix))
+transmission
+```
+
+This can be given as a parameter to the model by adding the argument `data = list(m = transmission)` when using the `new()` method on the `gen_age` generator.
+
+## Saving a model into a package
+
+If you want to distribute your model in an R package, rather than building it locally, you will want to use the `odin_dust_package()` function instead. This will write the transpiled C++ dust code into `src` and its R interface in `R/dust.R`. Package users are not required to regenerate the dust code, and their compiler will build the library when they install the package. You may also wish to add a file `R/zzz.R` to your package `mycoolpackage` with the following lines:
+
+```r
+##' @useDynLib mycoolpackage, .registration = TRUE
+NULL
+```
+
+to help automate the compilation of the model.
+
+# Running the SIR model with [`dust`](https://mrc-ide.github.io/dust/)
```

Comment:
IMO all the material above here can be dropped and/or moved to different vignettes. I think as much as possible this vignette should focus on the "I want to do inference" problem probably. We can have as many vignettes as needed and IME keeping them fairly focussed, while hard, is necessary to keep them clear

Reply:
Moving everything above the mcstate section to odin.dust